src/storage.rs

```diff
@@ -1332,44 +1345,56 @@ fn parse_imported_criteria(value: toml::Value) -> Option<CriteriaEntry> {
         .ok()
 }
 
-/// Parse an unparsed audit entry, validating and returning it.
-fn parse_imported_audit(valid_criteria: &[CriteriaName], value: toml::Value) -> Option<AuditEntry> {
-    let mut audit: AuditEntry = parse_from_value(value)
-        .map_err(|err| info!("imported audit parsing failed due to {err}"))
-        .ok()?;
-
+fn retain_only_known_criteria(
+    audit_criteria: &mut Vec<Spanned<CriteriaName>>,
+    valid_criteria: &[CriteriaName],
+) -> Result<(), String> {
     // Remove any unrecognized criteria to avoid later errors caused by being
     // unable to find criteria, and ignore the entry if it names no known
     // criteria.
-    audit
-        .criteria
-        .retain(|criteria_name| is_known_criteria(valid_criteria, criteria_name));
-    if audit.criteria.is_empty() {
-        info!("imported audit parsing failed due to no known criteria");
-        return None;
+    if audit_criteria
+        .iter()
+        .any(|criteria_name| is_known_criteria(valid_criteria, criteria_name))
+    {
+        audit_criteria.retain(|criteria_name| is_known_criteria(valid_criteria, criteria_name));
+        Ok(())
+    } else {
+        Err(format!(
+            "no known criteria: {}",
+            audit_criteria
+                .iter()
+                .map(ToString::to_string)
+                .collect::<Vec<_>>()
+                .join(", "),
+        ))
     }
```

Comment:
Rather than doing the split like this, it might be cleaner (and help with situations where some but not all criteria are discarded) to log incrementally for each criteria removed, like this:

```rust
// Could be more elegant with `Vec::extract_if`, however that is still unstable
audit.criteria.retain(|criteria_name| {
    if !is_known_criteria(valid_criteria, criteria_name) {
        info!("discarding unknown criteria in imported audit: {criteria_name}");
        return false;
    }
    true
});
if audit.criteria.is_empty() {
    info!("imported audit parsing failed due to no known criteria");
    return None;
}
```

Reply:
Done (I think).